packages/vitest/src/types/coverage.ts

```diff
@@ -15,41 +15,101 @@ export type CoverageReporter =
   | 'text-summary'
   | 'text'
 
-export interface C8Options {
+export type CoverageProviders = 'c8' | 'istanbul'
+
+export type CoverageOptions =
+  | NullCoverageOptions & { provider?: null }
+  | C8Options & { provider?: 'c8' }
+  | IstanbulOptions & { provider?: 'istanbul' }
+
+interface BaseCoverageOptions {
   /**
    * Enable coverage, pass `--coverage` to enable
    *
    * @default false
    */
   enabled?: boolean
-  /**
-   * Directory to write coverage report to
-   */
-  reportsDirectory?: string
+
   /**
    * Clean coverage before running tests
    *
    * @default true
    */
   clean?: boolean
+
   /**
    * Clean coverage report on watch rerun
    *
    * @default false
    */
   cleanOnRerun?: boolean
+
   /**
-   * Allow files from outside of your cwd.
-   *
-   * @default false
+   * Directory to write coverage report to
    */
-  allowExternal?: any
+  reportsDirectory?: string
+
   /**
    * Reporters
    *
    * @default 'text'
    */
   reporter?: Arrayable<CoverageReporter>
+
+}
+
+export interface NullCoverageOptions extends BaseCoverageOptions {
+  enabled: false
+}
+
+export interface IstanbulOptions extends BaseCoverageOptions {
+  /* Name of global coverage variable. (optional, default __coverage__) */
+  coverageVariable?: string
```

Comment:
This should not be public as users should not change it. Same for `coverageGlobal*` options.